### ISSUE TITLE
The mobilenet predict results are different

### ISSUE THREAD

**Issue by longchr123** [tensorflow/tensorflow#21124]
**Problem Description:**

I meet a very strange problem,the problem is that:
My job is train a pb file that can predict face's five key points,he network model is mobilenet v2,when I trained to get a small loss pb file, the loss is about 0.7,that's ok,no problem.But when I predict with trained pb file, there's a problem. When the input image shape is [100,224,224,3], get the output is [100,10],(ten means five key point coordinates),that's ok, the output is accurate,tes is right,but when I change input image shape[1,224,224,3](there's only one input image),the output [1,10] is different from privous result,the two cases of  input[0] is the same, but output[0] is different,  the input shape is  [100,224,224,3] more accurate, why? 

the exact coordinates of output[0]:
[38 46 71 54 62 70 26 81 56 88]

input shape [100,224,224,3], get input[0]:
[39.381622 45.988384 73.15548  55.10713  64.56505  70.48474  29.348461
 82.159615 53.589706 87.59885 ]

input shape [1,224,224,3],get input[0]:
[[40.17595  47.69543  85.54374  50.631523 64.47522  71.54203  36.73784
  85.07918  79.665985 87.313995]]

the exact coordinates of output[0]:
[39 46 88 35 76 65 51 80 94 70]

input shape [100,224,224,3], get input[0]:
[38.898895 45.90179  86.28276  34.750797 76.82557  65.86458  51.15553
 79.85125  95.181885 70.00042 ]

input shape [1,224,224,3],get input[0]:
[[39.191376 46.032986 88.266106 39.030933 72.28544  64.2173   49.861656
  80.540855 94.46614  74.561905]]

-------------------------------------

System information
Linux Ubuntu 16.04:
TensorFlow installed from source:
TensorFlow version 1.8.0:
Python version:3.6:
Bazel version 0.11.1:
GCC/Compiler version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.10) :
CUDA/cuDNN 9.0/7.0.5:
GPU 1060-6G:

**Comment by longchr123** [comment]
the batch size of the training is 100

**Comment by longchr123** [comment]
when input shape is [50,224,224,3],the accurate less [100,224,224,3],but more than [1,224,224,3]

**Comment by longchr123** [comment]
Especially on the side face,the error will be greater, like this face:

the exact coordinates of output[0]:
38 46 71 54 62 70 26 81 56 88

input shape [100,224,224,3], get input[0]:
[39.381622 45.988384 73.15548  55.10713  64.56505  70.48474  29.348461
 82.159615 53.589706 87.59885 ]

input shape [1,224,224,3],get input[0]:
[[40.17595  47.69543  85.54374  50.631523 64.47522  71.54203  36.73784
  85.07918  79.665985 87.313995]]

**Comment by atong01** [comment]
Forgetting to set your batch-norm layers to training=False mode might cause this. As in training mode batch-norm uses batch-wise statistics of mean and variance as estimates for dataset statistics which are clearly less accurate with smaller batches.

**Comment by longchr123** [comment]
**network demo**
`
weight_decay = 1e-4


def relu(x, name='relu6'):
    return tf.nn.relu6(x, name)


def batch_norm(x, momentum=0.9, epsilon=1e-5, train=True, name='bn'):
    return tf.layers.batch_normalization(x,
                      momentum=momentum,
                      epsilon=epsilon,
                      scale=True,
                      training=train,
                      name=name)

def conv2d(input_, output_dim, k_h, k_w, d_h, d_w, stddev=0.02, name='conv2d', bias=False):
    with tf.variable_scope(name):
        w = tf.get_variable('w', [k_h, k_w, input_.get_shape()[-1], output_dim],
              regularizer=tf.contrib.layers.l2_regularizer(weight_decay),
              initializer=tf.truncated_normal_initializer(stddev=stddev))
        conv = tf.nn.conv2d(input_, w, strides=[1, d_h, d_w, 1], padding='SAME')
        if bias:
            biases = tf.get_variable('bias', [output_dim], initializer=tf.constant_initializer(0.0))
            conv = tf.nn.bias_add(conv, biases)

        return conv

def conv2d_block(input, out_dim, k, s, is_train, name):
    with tf.name_scope(name), tf.variable_scope(name):
        net = conv2d(input, out_dim, k, k, s, s, name='conv2d')
        net = batch_norm(net, train=is_train, name='bn')
        net = relu(net)
        return net


def conv_1x1(input, output_dim, name, bias=False):
    with tf.name_scope(name):
        return conv2d(input, output_dim, 1,1,1,1, stddev=0.02, name=name, bias=bias)


def pwise_block(input, output_dim, is_train, name, bias=False):
    with tf.name_scope(name), tf.variable_scope(name):
        out=conv_1x1(input, output_dim, bias=bias, name='pwb')
        out=batch_norm(out, train=is_train, name='bn')
        out=relu(out)
        return out

def global_avg(x):
    with tf.name_scope('global_avg'):
        net=tf.layers.average_pooling2d(x, x.get_shape()[1:-1], 1)
        return net


def flatten(x):
    return tf.contrib.layers.flatten(x)


def mobilenetv2(inputs, num_output, is_train=True):
	with tf.variable_scope('mobilenetv2_128'):
		net = conv2d_block(inputs, 32, 3, 2, is_train, name='conv1_1')


        net = pwise_block(net, 160, is_train, name='conv7_1')

        net = global_avg(net)

        net = flatten(conv_1x1(net, num_output, name='logits'))
	return net

def loss(labels,predictions,scope='loss'):
    labels = tf.reshape(labels, shape=(-1, 2))
    predictions = tf.reshape(predictions, shape=(-1, 2))

    # 计算两矩阵对应点相减后的平方
    dis = tf.square(labels - predictions)

    # 在行的维度上相加
    dis2 = tf.reduce_sum(dis, axis=1)
    # 类型转换
    dis2 = tf.cast(dis2, dtype=tf.float32)

    # 求开方
    sqrt = tf.sqrt(dis2)

    # 求均值
    mean = tf.reduce_mean(sqrt,name=scope)

    return tf.maximum(mean,1e-5)
`

**test demo**
`

def read_image_cv(image_dir,sess):  # read image, bbx and landmarks

    image_names = os.listdir(image_dir)

    image_names = sorted(image_names)

    image_paths = [image_dir + name for name in image_names]

    image_list = []
    for image_path in image_paths:

        image = cv.imread(image_path)
        image = tf.reshape(image, [128, 128, 3])
        image = tf.cast(image, tf.float64)
        image = tf.image.per_image_standardization(image)

        x_test = image.eval(session=sess)

        image_list.append(x_test)
    return image_list

def predict_test_set_index(meta_path,checkpoint_path, test_image_dir, index):
    saver = tf.train.import_meta_graph(meta_path)
    saver.restore(sess, checkpoint_path)

    test_image_path = [test_image_dir + name for name in sorted(os.listdir(test_image_dir))]

    with tf.Session() as session:
        input = tf.get_default_graph().get_tensor_by_name("input:0")
        output = tf.get_default_graph().get_tensor_by_name("output:0")
        image = read_image_cv(test_image_dir, session)

        out = session.run(output, feed_dict={input: image})

    image_path = test_image_path[index]
    return out, image_path

def predict_test_one(meta_path,checkpoint_path, test_image_path):
    saver = tf.train.import_meta_graph(meta_path)
    saver.restore(sess, checkpoint_path)

    with tf.Session() as session:
        input = tf.get_default_graph().get_tensor_by_name("input:0")
        output = tf.get_default_graph().get_tensor_by_name("output:0")
        image = cv.imread(test_image_path)
        image = tf.reshape(image, [128, 128, 3])
        image = tf.cast(image, tf.float64)
        image = tf.image.per_image_standardization(image)

        x_test = image.eval(session=session)
        x_test = x_test[np.newaxis, :]

        out = session.run(output, feed_dict={input: x_test})
    return out

def test_one_img():
    image_name = '182121.jpg'
    output = predict_test_one(meta_path,checkpoint_path, TEST_IMAGINE_PATH+image_name)
    print(output)
# print(output)
#draw_landmarks.draw_graph(TEST_IMAGINE_PATH, output, graph_num)


def test_set_img():
    index = 0
    # there are 100 images in this directory(TEST_IMAGINE_PATH),the index 0 is 182121.jpg
    output,image_path= predict_test_set_index(meta_path,checkpoint_path, TEST_IMAGINE_PATH,index)
    print(output[index])

test_set_img()
test_one_img()
`

**Comment by longchr123** [comment]
**network demo**
`
weight_decay = 1e-4
def relu(x, name='relu6'):
    return tf.nn.relu6(x, name)

def batch_norm(x, momentum=0.9, epsilon=1e-5, train=True, name='bn'):
    return tf.layers.batch_normalization(x,
                      momentum=momentum,
                      epsilon=epsilon,
                      scale=True,
                      training=train,
                      name=name)

def conv2d(input_, output_dim, k_h, k_w, d_h, d_w, stddev=0.02, name='conv2d', bias=False):
    with tf.variable_scope(name):
        w = tf.get_variable('w', [k_h, k_w, input_.get_shape()[-1], output_dim],
              regularizer=tf.contrib.layers.l2_regularizer(weight_decay),
              initializer=tf.truncated_normal_initializer(stddev=stddev))
        conv = tf.nn.conv2d(input_, w, strides=[1, d_h, d_w, 1], padding='SAME')
        if bias:
            biases = tf.get_variable('bias', [output_dim], initializer=tf.constant_initializer(0.0))
            conv = tf.nn.bias_add(conv, biases)
        return conv
def conv2d_block(input, out_dim, k, s, is_train, name):
    with tf.name_scope(name), tf.variable_scope(name):
        net = conv2d(input, out_dim, k, k, s, s, name='conv2d')
        net = batch_norm(net, train=is_train, name='bn')
        net = relu(net)
        return net
def conv_1x1(input, output_dim, name, bias=False):
    with tf.name_scope(name):
        return conv2d(input, output_dim, 1,1,1,1, stddev=0.02, name=name, bias=bias)
def pwise_block(input, output_dim, is_train, name, bias=False):
    with tf.name_scope(name), tf.variable_scope(name):
        out=conv_1x1(input, output_dim, bias=bias, name='pwb')
        out=batch_norm(out, train=is_train, name='bn')
        out=relu(out)
        return out
def global_avg(x):
    with tf.name_scope('global_avg'):
        net=tf.layers.average_pooling2d(x, x.get_shape()[1:-1], 1)
        return net
def flatten(x):
    return tf.contrib.layers.flatten(x)
def mobilenetv2(inputs, num_output, is_train=True):
	with tf.variable_scope('mobilenetv2_128'):
		net = conv2d_block(inputs, 32, 3, 2, is_train, name='conv1_1')
                net = pwise_block(net, 160, is_train, name='conv7_1')
                net = global_avg(net)
                net = flatten(conv_1x1(net, num_output, name='logits'))
	return net
`

**test demo**
`
def read_image_cv(image_dir,sess):  # read image, bbx and landmarks
    image_names = os.listdir(image_dir)
    image_names = sorted(image_names)
    image_paths = [image_dir + name for name in image_names]
    image_list = []
    for image_path in image_paths:
        image = cv.imread(image_path)
        image = tf.reshape(image, [128, 128, 3])
        image = tf.cast(image, tf.float64)
        image = tf.image.per_image_standardization(image)
        x_test = image.eval(session=sess)
        image_list.append(x_test)
    return image_list
def predict_test_set_index(meta_path,checkpoint_path, test_image_dir, index):
    saver = tf.train.import_meta_graph(meta_path)
    saver.restore(sess, checkpoint_path)
    test_image_path = [test_image_dir + name for name in sorted(os.listdir(test_image_dir))]
    with tf.Session() as session:
        input = tf.get_default_graph().get_tensor_by_name("input:0")
        output = tf.get_default_graph().get_tensor_by_name("output:0")
        image = read_image_cv(test_image_dir, session)
        out = session.run(output, feed_dict={input: image})
    image_path = test_image_path[index]
    return out, image_path
def predict_test_one(meta_path,checkpoint_path, test_image_path):
    saver = tf.train.import_meta_graph(meta_path)
    saver.restore(sess, checkpoint_path)
    with tf.Session() as session:
        input = tf.get_default_graph().get_tensor_by_name("input:0")
        output = tf.get_default_graph().get_tensor_by_name("output:0")
        image = cv.imread(test_image_path)
        image = tf.reshape(image, [128, 128, 3])
        image = tf.cast(image, tf.float64)
        image = tf.image.per_image_standardization(image)
        x_test = image.eval(session=session)
        x_test = x_test[np.newaxis, :]
        out = session.run(output, feed_dict={input: x_test})
    return out
def test_one_img():
    image_name = '182121.jpg'
    output = predict_test_one(meta_path,checkpoint_path, TEST_IMAGINE_PATH+image_name)
    print(output)
def test_set_img():
    index = 0
    # there are 100 images in this directory(TEST_IMAGINE_PATH),the index 0 is 182121.jpg
    output,image_path= predict_test_set_index(meta_path,checkpoint_path, TEST_IMAGINE_PATH,index)
    print(output[index])

test_set_img()
test_one_img()
`

**Comment by longchr123** [comment]
when i remove batch_norm function, the problem is still there

**Comment by longchr123** [comment]
I sovled this preblom,this is because BN based on batch_size,when batch_size=1,the error will be enlarged,so can't use bn,we should use layer norm , instance norm or group norm, link:https://blog.csdn.net/qq_25737169/article/details/79713886
